### ISSUE TITLE
Output shape argument description in tf.keras.layers.MultiHeadAttention possibly wrong.

### ISSUE THREAD

**Issue by CORRELU** [keras-team/keras#17481]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**URL(s) with the issue:**.

[Link to MultiHeadAttention docs: https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention/
](https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention)

**Description of issue (what needs to be changed):**.
Output shape argument description currently says the following: _The expected shape of an output tensor, besides the batch and sequence dims. If not specified, projects back to the **key** feature dim._ I believe it should say:  _The expected shape of an output tensor, besides the batch and sequence dims. If not specified, projects back to the **input** feature dim._  

**Correct links**.
Yes

**Parameters defined**.

Yes

**Returns defined**.

Yes

**Raises listed and defined**

Yes

**Usage example**

This can be confirmed with the following code:
```
import tensorflow as tf
mha = tf.keras.layers.MultiHeadAttention(key_dim=2, num_heads=8)
input = tf.keras.Input(shape=[256, 18])
attention = mha(input, input, input)
print(mha.weights[6].shape[-1])
```
The code will output 18 (the input feature dimension) rather than the key feature dimension (in this case, 2).

**Request visuals, if applicable**.

No and no.

**Submit a pull request?**.

Yes

**Comment by tilakrayal** [comment]
@CORRELU ,
When using MultiHeadAttention inside a custom layer, the custom layer must implement its own build() method and call MultiHeadAttention's _build_from_signature() there. This enables weights to be restored correctly when the model is loaded.

Also When I was trying to execute the given code in the alternative approach by providing the **(mha.weights[5].shape[-1])**, the output was 2 which is **key_dim** as mentioned in the document. 

```
import tensorflow as tf
mha = tf.keras.layers.MultiHeadAttention(key_dim=2, num_heads=8)
input = tf.keras.Input(shape=[256, 18])
attention = mha(input, input, input)
print(mha.weights[5].shape[-1]) 
```

Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/1a76f8cb8a3798f006145df14b96a4b0/untitled911.ipynb) for the reference. Thank you!

**Comment by CORRELU** [comment]
@tilakrayal 
Thank you for your response. Index 5 in mha.weights corresponds to the value bias not the key , it becomes apparent when printing out the name :`mha.weights[5].name`. 

Let me rephrase my question. It says that the output dimension will correspond to the key dimension if no argument is given, however for k_dim=2, mha.output_shape[-1] clearly outputs the dimension of the input not the key. After thinking about it a little more, I realise the documentation may be refering to the key before it its transformed using the weights and bias, in which case the key dimension is indeed 18. You can close the issue now.

**Comment by tilakrayal** [comment]
@CORRELU,
Yes, As mentioned before it is transformed using the weights and bias. it is referring to the key.
Also the expected shape of an output tensor, besides the batch and sequence dims. If not specified, projects back to the key feature dim. 

Glad the issue is resolved for you, Could you please feel free to move this issue to closed status. Thank you!

**Comment by tilakrayal** [comment]
@CORRELU,
Yes, As mentioned before it is transformed using the weights and bias. it is refering to the key.
Also the expected shape of an output tensor, besides the batch and sequence dims. If not specified, projects back to the key feature dim. 

Glad the issue is resolved for you, Could you please feel free to move this issue to closed status. Thank you!
